server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java

```diff
@@ -1005,19 +1012,37 @@ private void setUpdateTablet(UpdateSession us, KeyExtent keyExtent) {
         updateMetrics.addPermissionErrors(0);
         return;
       }
+
     }
 
     @Override
     public void applyUpdates(TInfo tinfo, long updateID, TKeyExtent tkeyExtent,
-        List<TMutation> tmutations) {
+        List<TMutation> tmutations) throws TException {
       UpdateSession us = (UpdateSession) sessionManager.reserveSession(updateID);
       if (us == null) {
         return;
       }
 
       boolean reserved = true;
+      boolean allowWriteThreadSemaphore = false;
       try {
         KeyExtent keyExtent = new KeyExtent(tkeyExtent);
+        maxThreads = getServerConfig().getConfiguration().getCount(Property.TSERV_MAX_WRITETHREADS);
+        maxThreadPermits = maxThreads == 0 ? Integer.MAX_VALUE : maxThreads;
+        sem = new Semaphore(maxThreadPermits);
+
+        if (TabletType.type(keyExtent) == TabletType.USER) {
+          if (!sem.tryAcquire()) {
+            us.failures.put(keyExtent, 0L);
+            updateMetrics.addUnknownTabletErrors(0);
+            log.error("Mutation failed - No more threads available for mutations at " + new Date());
```

Comment:
Seems like this should throw an exception.

Reply:
I now throw a TException here so we can have feedback when a mutation fails.